### PR TITLE
Replace curse scrolls with curse item ability.

### PR DESCRIPTION
**Pull request by alexjurkiewicz**
This zero-piety ability of Ashenzari curses an item in exchange for one
scroll of remove curse.

This puts more pressure on ?rc scrolls for Ash worshippers, but reduces
the risk of getting the "wrong" curse scrolls, especially in early game.

Removes:
* Praying on ?rc to give ?curse weapon/armour/jewellery
* Curse weapon/armour/jewellery scrolls

### CODE REVIEW

File: crawl-ref/source/ability.cc [hunk 1]
```diff
@@ -2794,6 +2800,25 @@ static spret_type _do_ability(const ability_def& abil, bool fail)
...
+                dec_inv_item_quantity(i, 1);
+                success = ashenzari_curse_item();
+                break;
+            }
+        }
+        if (!success)
+            mpr("You need a scroll of remove curse to do this.");
+        break;
+    }
+
```
Comment (alexjurkiewicz) [inline crawl-ref/source/ability.cc]: Should the check logic be moved elsewhere? I'm not sure where it's meant to go...

File: crawl-ref/source/godabil.cc [hunk 2]
```diff
@@ -3792,6 +3793,32 @@ bool ashenzari_end_transfer(bool finished, bool force)
     return true;
 }
 
+bool ashenzari_curse_item()
+{
+    while (1)
+    {
+        int item_slot = prompt_invent_item("Curse which item?", MT_INVLIST,
```
Comment (alexjurkiewicz) [inline crawl-ref/source/godabil.cc]: Should this menu mention you can press ESC to cancel?

File: crawl-ref/source/ability.cc [hunk 3]
```diff
@@ -2794,6 +2800,25 @@ static spret_type _do_ability(const ability_def& abil, bool fail)
...
+    case ABIL_ASHENZARI_CURSE:
+    {
+        fail_check();
+        bool success = false;
+        for (int i = 0; i < ENDOFPACK; ++i)
+        {
+            item_def item = you.inv[i];
+            if (item.base_type == OBJ_SCROLLS && item.sub_type == SCR_REMOVE_CURSE)
+            {
+                dec_inv_item_quantity(i, 1);
```
Comment (alexjurkiewicz) [inline crawl-ref/source/ability.cc]: This should only happen if the next line succeeds

File: crawl-ref/source/godabil.cc [hunk 4]
```diff
@@ -3792,6 +3793,32 @@ bool ashenzari_end_transfer(bool finished, bool force)
...
+                                           OSEL_CURSABLE,
+                                           true, true, false);
+        if (prompt_failed(item_slot))
+            return false;
+
+        item_def& item(you.inv[item_slot]);
+
+        ASSERT(!item.cursed());
+        if (!item_is_equipped(item))
```
Comment (alexjurkiewicz) [inline crawl-ref/source/godabil.cc]: Change the OSEL to only be equipped uncursed items and change this line to an ASSERT

File: crawl-ref/source/invent.h [hunk 5]
```diff
@@ -35,6 +35,7 @@ enum object_selector
     OSEL_ENCHANTABLE_WEAPON      = -17,
     OSEL_BLESSABLE_WEAPON        = -18,
     OSEL_SUPERCHARGE             = -19,
+    OSEL_CURSABLE                = -20,
```
Comment (alexjurkiewicz) [inline crawl-ref/source/invent.h]: change to OSEL_EQUIPPED_CURSABLE

**Comment by alexjurkiewicz** [comment]
This PR needs a tile, and also I'm not 100% sure what happens with save transferred games

File: crawl-ref/source/ability.cc (hunk 3 above)
Comment (alexjurkiewicz) [inline crawl-ref/source/ability.cc]: done

File: crawl-ref/source/ability.cc (hunk 1 above)
Comment (alexjurkiewicz) [inline crawl-ref/source/ability.cc]: done

File: crawl-ref/source/godabil.cc (hunk 4 above)
Comment (alexjurkiewicz) [inline crawl-ref/source/godabil.cc]: OSEL was changed to be "cursable and not known-cursed".

File: crawl-ref/source/enum.h [hunk 6]
```diff
@@ -370,6 +370,7 @@ enum ability_type
     ABIL_ASHENZARI_SCRYING = 1160,
     ABIL_ASHENZARI_TRANSFER_KNOWLEDGE,
     ABIL_ASHENZARI_END_TRANSFER,
+    ABIL_ASHENZARI_CURSE,
```
Comment (alexjurkiewicz) [inline crawl-ref/source/enum.h]: is this ordering OK? I added it at the end for save-compat reasons, but it's out of order with the ability unlocks

Comment (wheals) [inline crawl-ref/source/enum.h]: Yeah, it's fine.

File: crawl-ref/source/ability.cc [hunk 7]
```diff
@@ -2794,6 +2800,32 @@ static spret_type _do_ability(const ability_def& abil, bool fail)
...
+    case ABIL_ASHENZARI_CURSE:
+    {
+        fail_check();
+        auto iter = find_if(
+                            begin(you.inv),
+                            end(you.inv),
+                            [] (const item_def &it) {
+                                return it.is_type(OBJ_SCROLLS, SCR_REMOVE_CURSE);
+                            });
+        if (iter != end(you.inv)) {
```
Comment (wheals) [inline crawl-ref/source/ability.cc]: braces on new line (also on the `else` below)

File: crawl-ref/source/itemprop.cc [hunk 8]
```diff
@@ -733,6 +733,17 @@ bool item_known_cursed(const item_def &item)
     return _full_ident_mask(item) & ISFLAG_KNOW_CURSE
            && item_ident(item, ISFLAG_KNOW_CURSE) && item.cursed();
 }
+bool item_is_cursable(const item_def &item, bool ignore_holy_wrath)
```
Comment (wheals) [inline crawl-ref/source/itemprop.cc]: newline between functions

File: crawl-ref/source/itemprop.cc [hunk 9]
```diff
@@ -733,6 +733,17 @@ bool item_known_cursed(const item_def &item)
...
+{
+    if (!item_type_has_curses(item.base_type))
+        return false;
+    if (item_known_cursed(item))
+        return false;
+    if (!ignore_holy_wrath && item.base_type == OBJ_WEAPONS
+        && get_weapon_brand(item) == SPWPN_HOLY_WRATH)
+        return false;
```
Comment (wheals) [inline crawl-ref/source/itemprop.cc]: braces after multi-line if

**Comment by wheals** [comment]
* It should be checked whether this can leak whether a weapon is holy wrath
* The curse scrolls should be added to `removed_items` and all mentions surrounded with `TAG_MAJOR_VERSION` checks or removed.

**Comment by alexjurkiewicz** [comment]
Would added curse scrolls to removed_items remove them from existing games?
I feel that would screw over current Ash games...

**Comment by wheals** [comment]
No, it would just make them not show up on `\` or matter for last-subtype ID: this is already the case, but it's done differently right now. There are a few other things it would affect, but at any rate they would still exist and be usable.

**Comment by alexjurkiewicz** [comment]
> It should be checked whether this can leak whether a weapon is holy wrath

The code used could if a non-Ash worshipper ever gets access, so I added that to the function documentation.

> The curse scrolls should be added to removed_items and all mentions surrounded with TAG_MAJOR_VERSION checks or removed.

done

**Comment by alexjurkiewicz** [comment]
A tiles icon for the ability would be nice, by the way.

File: crawl-ref/source/ability.cc [hunk 10]
```diff
@@ -79,30 +79,30 @@
 
 enum class abflag
 {
-    NONE           = 0x00000000,
```
Comment (amalloy) [inline crawl-ref/source/ability.cc]: Changes to whitespace in otherwise-unchanged code should really be in a separate commit.

Comment (wheals) [inline crawl-ref/source/ability.cc]: It's not unrelated; it's keeping them all aligned.

File: crawl-ref/source/itemname.cc [hunk 11]
```diff
@@ -2604,6 +2599,7 @@ void check_item_knowledge(bool unknown_items)
             if (item_type_removed(i, j))
                 continue;
 
+#if TAG_MAJOR_VERSION == 34
             // Curse scrolls are only created by Ashenzari.
```
Comment (wheals) [inline crawl-ref/source/itemname.cc]: This is redundant with the check right above it now.

File: crawl-ref/source/ng-setup.cc [hunk 12]
```diff
@@ -372,10 +372,12 @@ static void _give_basic_knowledge()
     // Recognisable by appearance.
     you.type_ids[OBJ_POTIONS][POT_BLOOD] = true;
 
+#if TAG_MAJOR_VERSION == 34
```
Comment (wheals) [inline crawl-ref/source/ng-setup.cc]: This is unnecessary now; see the comment right below.

File: crawl-ref/source/ability.cc [hunk 13]
```diff
@@ -2787,6 +2799,36 @@ static spret_type _do_ability(const ability_def& abil, bool fail)
...
+        fail_check();
+        auto iter = find_if(
+                            begin(you.inv),
+                            end(you.inv),
+                            [] (const item_def &it)
+                            {
+                                return it.is_type(OBJ_SCROLLS, SCR_REMOVE_CURSE);
+                            }
+                           );
+        if (iter != end(you.inv)) {
```
Comment (wheals) [inline crawl-ref/source/ability.cc]: separate line

**Comment by alexjurkiewicz** [comment]
Thanks for your reviewing here wheals :-)
